src/evo/providertx.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2017 The Dash Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "providertx.h"
+#include "specialtx.h"
+
+#include "hash.h"
+#include "clientversion.h"
+#include "streams.h"
+#include "messagesigner.h"
+#include "chainparams.h"
+#include "validation.h"
+#include "univalue.h"
+#include "core_io.h"
+#include "script/standard.h"
+#include "base58.h"
+
+template <typename ProTx>
+static bool CheckService(const uint256& proTxHash, const ProTx& proTx, const CBlockIndex* pindexPrev, CValidationState& state)
+{
+    if (proTx.nProtocolVersion < MIN_PROTX_PROTO_VERSION || proTx.nProtocolVersion > PROTOCOL_VERSION)
+        return state.DoS(10, false, REJECT_INVALID, "bad-protx-proto-version");
+
+    if (!proTx.addr.IsValid())
+        return state.DoS(10, false, REJECT_INVALID, "bad-protx-addr");
+    if (Params().NetworkIDString() != CBaseChainParams::REGTEST && !proTx.addr.IsRoutable())
+        return state.DoS(10, false, REJECT_INVALID, "bad-protx-addr");
+
+    if (!proTx.addr.IsIPv4())
+        return state.DoS(10, false, REJECT_INVALID, "bad-protx-addr");
+
+    return true;
+}
+
+template <typename ProTx>
+static bool CheckInputsHashAndSig(const CTransaction &tx, const ProTx& proTx, const CKeyID &keyID, CValidationState& state)
+{
+    uint256 inputsHash = CalcTxInputsHash(tx);
+    if (inputsHash != proTx.inputsHash)
+        return state.DoS(100, false, REJECT_INVALID, "bad-protx-inputs-hash");
+
+    std::string strError;
+    if (!CHashSigner::VerifyHash(::SerializeHash(proTx), keyID, proTx.vchSig, strError))
+        return state.DoS(100, false, REJECT_INVALID, "bad-protx-sig", false, strError);
+
+    return true;
+}
+
+bool CheckProRegTx(const CTransaction& tx, const CBlockIndex* pindexPrev, CValidationState& state)
+{
+    AssertLockHeld(cs_main);
+
+    CProRegTx ptx;
+    if (!GetTxPayload(tx, ptx))
+        return state.DoS(100, false, REJECT_INVALID, "bad-tx-payload");
+
+    if (ptx.nVersion != CProRegTx::CURRENT_VERSION)
```

Comment:
Whenever we bump CURRENT_VERSION, we'll have to use some kind of deployment logic...either BIP9 or a spork to activate it. So, new versions would only be allowed after it's ensured most of the network has upgraded.

I fear that allowing `ptx.nVersion >= CProRegTx::CURRENT_VERSION` would allow people to put whatever version they want on-chain, making it later hard to do real version bumps.

Reply:
By `ptx.nVersion > CProRegTx::CURRENT_VERSION` I meant that we would reject any protx with version _above_ currently known one while still accepting/processing the old ones. The logic for different/additional checks for new version protxes would then follow this one.

Reply:
Ah now I understand. I'll change it to `ptx.nVersion > CProRegTx::CURRENT_VERSION` then. In the case of bumping the version, the thing with the deployment logic applies. Or do I still miss something?

Reply:
No, that's it - adding this fix is all I wanted for now (assuming that deployment logic for new version is smth to be added whenever we decide to actually bump it) :)